### DISCUSSION TITLE
Mitmproxy sends no certificate to client

### DISCUSSION THREAD

**Discussion by Anonymous-Coward** [mitmproxy/mitmproxy#7424] [Q&A]
I have a Kubernetes cluster deployed localy with [kind](https://kind.sigs.k8s.io/), and some applications inside the cluster, using [nginx](https://github.com/kubernetes/ingress-nginx) for ingress. Without mitmproxy, everything works fine, and I can call the applications via https. The nginx ingress controller uses self-signed certificates.

If I follow the instructions from [here](https://docs.mitmproxy.org/stable/howto-transparent/#work-around-to-redirect-traffic-originating-from-the-machine-itself) and try to do the same requests, using curl, I get:

```
curl: (35) OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to localhost:443 
```

I also get an error when trying to open a URL of an application in a browser.

If I inspect the TLS session negotiation using `openssl s_client localhost:443` when mitmproxy is not running, I get something like this at the very beginning:

```
CONNECTED(00000003)
---
Certificate chain
 0 s:O=Acme Co, CN=Kubernetes Ingress Controller Fake Certificate
   i:O=Acme Co, CN=Kubernetes Ingress Controller Fake Certificate
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Dec 27 16:47:18 2024 GMT; NotAfter: Dec 27 16:47:18 2025 GMT
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIDcDCCAligAwIBAgIRALZtLMr5/R3YqBfi2KR3mN0wDQYJKoZIhvcNAQELBQAw
SzEQMA4GA1UEChMHQWNtZSBDbzE3MDUGA1UEAxMuS3ViZXJuZXRlcyBJbmdyZXNz
IENvbnRyb2xsZXIgRmFrZSBDZXJ0aWZpY2F0ZTAeFw0yNDEyMjcxNjQ3MThaFw0y
...
```

If I try to do the same when mitmproxy is running, I get this instead:

```
CONNECTED(00000003)
---
no peer certificate available
---
No client certificate CA names sent
---
...
```

`iptables -t nat -L OUTPUT` says:

```
Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
REDIRECT   tcp  --  anywhere             anywhere             ! owner UID match mitmproxy tcp dpt:https redir ports 8080
REDIRECT   tcp  --  anywhere             anywhere             ! owner UID match mitmproxy tcp dpt:http redir ports 8080
```

when the rules for proxying through mitmproxy are enabled and returns an empty list after I delete them.

Am I doing something wrong or is this a bug? Is there more information that would be needed to be provided in order to get an answer?

**Comment by Anonymous-Coward** [comment]
It's been three weeks. Does nobody have an answer?

**Comment by mhils** [comment]
This doesn't look like a mitmproxy bug, but a bug in your config. I would recommend trying out local capture mode instead of fiddling with iptables (https://mitmproxy.org/posts/local-capture/linux/).

**Reply by Anonymous-Coward** [reply]
@mhils Thank you - this worked for me, after upgrading from 11.0.2 to 11.1.0 0 - with 11.0.2 I was getting an error when trying to start mitmproxy:

```
mitmproxy --mode local:curl --showhost --ssl-insecure --mode transparent
```
However, the limitation that with `--mode local` mitmproxy can only capture outgoing connections, is pretty severe when you try to debug an ensemble of services communicating over https all of which are running locally - exactly my case.

The root cause of my problem is that with a self-signed certificate on the server side mitmproxy doesn't generate a certificate for the client. If mitmproxy would generate a certificate for the client when the server certificate is self-signed, I'd expect the approach I used initially to work perfectly. Is it intended behavior that mitmproxy doesn't send a certificate to the client when the server uses a self-signed certificate?

The approach I was using is documented on [the official mitmproxy documentation site](https://docs.mitmproxy.org/stable/howto-transparent/#work-around-to-redirect-traffic-originating-from-the-machine-itself). If this approach doesn't work that page should be removed, IMO. Also, `mitmproxy --help` makes no mention of `--mode local`. This, I believe, is also a bug, albeit a small and less important one. I can create tickets for both problems. Should I?

**Reply by mhils** [reply]
> The root cause of my problem is that with a self-signed certificate on the server side mitmproxy doesn't generate a certificate for the client. If mitmproxy would generate a certificate for the client when the server certificate is self-signed, I'd expect the approach I used initially to work perfectly. Is it intended behavior that mitmproxy doesn't send a certificate to the client when the server uses a self-signed certificate?

This sounds like a bug in your setup again. Extraordinary claims (mitmproxy does not send a certificate) require extraordinary evidence (e.g. clean packet captures).

> The approach I was using is documented on [the official mitmproxy documentation site](https://docs.mitmproxy.org/stable/howto-transparent/#work-around-to-redirect-traffic-originating-from-the-machine-itself). If this approach doesn't work that page should be removed, IMO. Also, `mitmproxy --help` makes no mention of `--mode local`. This, I believe, is also a bug, albeit a small and less important one. I can create tickets for both problems. Should I?

The instructions work for plenty of folks, but iptables is always fiddly and easy to get wrong. But `--mode local` is our attempt at providing an easier UX for this. No docs yet because it is still very new (https://mitmproxy.org/posts/local-capture/linux/), documentation work is tracked in https://github.com/mitmproxy/mitmproxy/issues/6531 and will come in a few weeks.

**Reply by Anonymous-Coward** [reply]
I am bit busy right now - wanted to do this last weekend but didn't get to it. Latest next weekend, however, I will do packet captures and attach them here. I did, however, describe above, in the initial question, what `openssl s_client` says. That's where I got the idea that mitmproxy does not send a certificate to the client.

Should I capture packets just for the traffic created by `openssl s_client`, to keep the capture files smaller, or should I perform a full request over https?

**Reply by Anonymous-Coward** [reply]
Just to let you know - today I did have the time to dig deeper into this, and now I believe it's a problem created by the software-defined networking of [kind](https://github.com/kubernetes-sigs/kind). I asked [a question](https://github.com/kubernetes-sigs/kind/issues/3867) there, with more details of what I've found, if you care to follow it. But given that now mitmproxy provides a way to capture and decrypt local traffic without manipulating the routing tables and that kind's software-defined networking is meant to be kept as simple as possible, I don't expect much interest in either project digging deeper into the issue. I won't bug you further with this issue, so feel free to close it if you don't want to spend more time on it. I did record some packet dumps, but given what I've found I don't know if they're relevant anymore.
